datafusion/functions-aggregate-common/src/aggregate/groups_accumulator/accumulate.rs

```diff
@@ -212,7 +229,66 @@ impl NullState {
     ///
     /// resets the internal state appropriately
     pub fn build(&mut self, emit_to: EmitTo) -> NullBuffer {
-        let nulls: BooleanBuffer = self.seen_values.finish();
+        self.seen_values.emit(emit_to)
+    }
+}
+
+/// Structure marking if accumulating groups are seen at least one
+pub trait SeenValues: Default + Debug + Send {
+    fn resize(&mut self, total_num_groups: usize, default_value: bool);
+
+    fn set_bit(&mut self, block_id: u32, block_offset: u64, value: bool);
+
+    fn emit(&mut self, emit_to: EmitTo) -> NullBuffer;
+
+    fn capacity(&self) -> usize;
+}
+
+/// [`SeenValues`] for `flat groups input`
+///
+/// The `flat groups input` are organized like:
+///
+/// ```text
+///     row_0 group_index_0
+///     row_1 group_index_1
+///     row_2 group_index_2
+///     ...
+///     row_n group_index_n     
+/// ```
+///
+/// If `row_x group_index_x` is not filtered(`group_index_x` is seen)
+/// `seen_values[group_index_x]` will be set to `true`.
+///
+/// For `set_bit(block_id, block_offset, value)`, `block_id` is unused,
```

Comment:
Something I didn't see documented anywhere was what `block_id` and `block_offset` meant -- maybe we could add something on the HashAggregateStream 🤔

Reply:
Addressed.
I add related comments in `GroupsAccumulator::supports_blocked_groups` and `GroupValues::supports_blocked_groups`, also link it in `HashAggregateStream`.